plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java

```diff
@@ -0,0 +1,90 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.runtime;
+
+import org.jkiss.code.NotNull;
+import org.jkiss.dbeaver.Log;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class ThreadExceptionsContext {
+
+    private static final Map<Thread, List<Exception>> THREAD_TO_EXCEPTIONS = new ConcurrentHashMap<>();
+    private static final Map<Thread, Thread> THREAD_BINDING_MAP = new ConcurrentHashMap<>();
+    private static final Log log = Log.getLog(ThreadExceptionsContext.class);
+
+    public static List<Exception> getListExceptionsForCurrentThread() {
+        Thread threadForRegistry = getThreadForRegistry();
+        List<Exception> exceptionList = THREAD_TO_EXCEPTIONS.get(threadForRegistry);
+        if (exceptionList == null) {
+            return new ArrayList<>();
+        }
+        return new ArrayList<>(exceptionList);
+    }
+
+    public static Set<Exception> getSetExceptionsForCurrentThread() {
+        return new HashSet<>(getListExceptionsForCurrentThread());
+    }
+
+    public static void registerExceptionForCurrentThread(Exception exception) {
+        Thread threadForRegistry = getThreadForRegistry();
+        List<Exception> exceptionList = THREAD_TO_EXCEPTIONS.get(threadForRegistry);
+        if (exceptionList == null) {
+            synchronized (threadForRegistry) {
+                exceptionList = THREAD_TO_EXCEPTIONS.get(threadForRegistry);
+                if (exceptionList == null) {
+                    exceptionList = new CopyOnWriteArrayList<>();
+                    exceptionList.add(exception);
+                    THREAD_TO_EXCEPTIONS.put(threadForRegistry, exceptionList);
+                    return;
+                }
+            }
+        }
+        exceptionList.add(exception);
+    }
+
+    /**
+     * If the thread which you are registering is a different thread that gonna read exception list
+     *
+     * @param callerThread a key where you will be able to find exceptions for current thread
+     */
+    public static void bindCurrentContextFor(Thread callerThread) {
+        Thread currentThread = Thread.currentThread();
+        //todo if it will need, there can be used a set for threads
+        if (THREAD_BINDING_MAP.get(currentThread) != null) {
+            log.warn(String.format("Thread %s is already in the binding map for ThreadExceptionContext. Overriding...",
```

Comment:
Do we really need to log it? User might see his message, but it's hard to understand, what this message is about, without context of this code

Reply:
I suppose there could be a concurrency issue that we can't see right now, so I prefer to see any such bugs related to concurrency in the log. I'm thinking of the log for developers, not for users, isn't it?

Reply:
Yes, but sometimes users read logs, if some bug occurred. 
If it's an exclusive situation, then I'm fine with logging it. I thought it will be called each time.